host/frontend/host-orchestrator/operation.go

```diff
@@ -20,83 +20,126 @@ import (
 	"sync"
 )
 
-type NewOperationError string
+type NotFoundOperationError string
 
-func (s NewOperationError) Error() string {
-	return fmt.Sprintf("failure creating a new operation: %s", string(s))
+func (s NotFoundOperationError) Error() string {
+	return fmt.Sprintf("operation not found: %s", string(s))
 }
 
-type OperationData struct {
+type Operation struct {
 	Name   string
 	Done   bool
-	Result OperationResultData
+	Result OperationResult
 }
 
-type OperationResultData struct {
-	Error OperationErrorData
-	OK    OperationOKData
+type OperationResult struct {
+	Error OperationResultError
 }
 
-type OperationErrorData struct {
+type OperationResultError struct {
 	ErrorMsg string
 }
 
-type OperationOKData struct{}
-
 type OperationManager interface {
-	NewOperation() (OperationData, error)
+	New() Operation
+
+	Get(string) (Operation, error)
 
-	GetOperation(string) (OperationData, bool)
+	Complete(string, OperationResult)
+
+	// TODO(b/231319087) This should handle timeout.
+	Wait(string)
+}
 
-	CompleteOperation(string, OperationResultData)
+type mapOMOperationEntry struct {
+	data  Operation
+	mutex sync.RWMutex
+	wait  sync.WaitGroup
 }
 
 type MapOM struct {
 	uuidFactory func() string
 	mutex       sync.RWMutex
-	operations  map[string]OperationData
+	operations  map[string]*mapOMOperationEntry
 }
 
 func NewMapOM(uuidFactory func() string) *MapOM {
 	return &MapOM{
 		uuidFactory: uuidFactory,
 		mutex:       sync.RWMutex{},
-		operations:  make(map[string]OperationData),
+		operations:  make(map[string]*mapOMOperationEntry),
 	}
 }
 
-func (m *MapOM) NewOperation() (OperationData, error) {
+const mapOMNewUUIDRetryLimit = 100
+
+func (m *MapOM) New() Operation {
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
-	name := m.uuidFactory()
-	if _, ok := m.operations[name]; ok {
-		return OperationData{}, NewOperationError("new operation name already exists")
+	name, retryCount := m.uuidFactory(), 0
+	_, found := m.operations[name]
+	for found && retryCount < mapOMNewUUIDRetryLimit {
+		name, retryCount = m.uuidFactory(), retryCount+1
+		_, found = m.operations[name]
+	}
+	if retryCount == mapOMNewUUIDRetryLimit {
+		panic("uuid retry limit reached")
+	}
+	wg := sync.WaitGroup{}
```

Comment:
https://pkg.go.dev/sync#WaitGroup says "A WaitGroup must not be copied after first use". It's not clear what first use means (does Add(1) count as first use?), but it's probably safer to just instantiate it as part of the struct directly and increment it after.

Reply:
great catch, done.